### PR TITLE
fix terser webpack plugin compatibility

### LINKED ISSUE #426: Add support for terser-webpack-plugin (webpack@4.26.0+)
Webpack 4.26.0 switched from using `uglifyjs-webpack-plugin` to `terser-webpack-plugin`. Since OfflinePlugin seems to partly piggyback off webpack's minifier but also contains uglifyjs-specific code this is now throwing up the following error:

```
OfflinePlugin: uglifyjs-webpack-plugin is required to preform a minification
```

Messed around a little bit myself but adding terser support while maintaining uglifyjs support for `< webpack@4.26.0` involves a bit more than updating the import, unfortunately.

Those running into this issue can meanwhile explicitly add `uglifyjs-webpack-plugin` to the dev dependencies until this is resolved.

### PR DESCRIPTION
**Pull request by Bobgy**
closes #426

### CODE REVIEW

**Comment by Bobgy** [comment]
I didn't update tests, but test is pulling in latest webpack by default. Seeing CIs passing should already prove the new version works with latest webpack without changing fixtures a little bit.

**Comment by Bobgy** [comment]
New behavior:
1. when terser-webpack-plugin exists, it will use it instead of uglify-webpack-plugin
2. when using terser-webpack-plugin, it will pass the same options from uglify-webpack-plugin to it

File: tests/legacy/fixtures/sw-minify-false/webpack.config.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
-var UglifyJsPlugin = require(__ROOT__ + '/lib/misc/get-uglify-plugin');
+var UglifyUtils = require(__ROOT__ + '/lib/misc/get-uglify-plugin');
```
Comment (GGAlanSmithee) [inline tests/legacy/fixtures/sw-minify-false/webpack.config.js]: var makeUglifyJsPlugin = require(__ROOT__ + '/lib/misc/get-uglify-plugin').makeUglifyJsPlugin;

File: tests/legacy/fixtures/sw-minify-auto/webpack.config.js [hunk 2]
```diff
@@ -1,4 +1,5 @@
-var UglifyJsPlugin = require(__ROOT__ + '/lib/misc/get-uglify-plugin');
+var UglifyUtils = require(__ROOT__ + '/lib/misc/get-uglify-plugin');
```
Comment (GGAlanSmithee) [inline tests/legacy/fixtures/sw-minify-auto/webpack.config.js]: var makeUglifyJsPlugin = require(ROOT + '/lib/misc/get-uglify-plugin').makeUglifyJsPlugin;

File: src/service-worker.js [hunk 3]
```diff
@@ -82,28 +82,28 @@ export default class ServiceWorker {
...
         test: new RegExp(name),
-        uglifyOptions: uglifyOptions
+        uglifyOptions,
       };
 
-      new UglifyJsPlugin(options).apply(childCompiler);
+      makeUglifyJsPlugin(options).apply(childCompiler);
     } else if (
-      (this.minify !== false || optimization.minimize) && UglifyJsPlugin
+      (this.minify !== false || optimization.minimize) && makeUglifyJsPlugin
```
Comment (GGAlanSmithee) [inline src/service-worker.js]: `makeUglifyJsPlugin` should always be `true` here

File: src/service-worker.js [hunk 4]
```diff
@@ -82,28 +82,28 @@ export default class ServiceWorker {
     };
 
     if (this.minify === true) {
-      if (!UglifyJsPlugin) {
-        throw new Error('OfflinePlugin: uglifyjs-webpack-plugin is required to preform a minification')
+      if (!makeUglifyJsPlugin) {
```
Comment (GGAlanSmithee) [inline src/service-worker.js]: I think `makeUglifyJsPlugin !== null` would be more explicit

File: src/misc/get-uglify-plugin.js [hunk 5]
```diff
@@ -1,5 +1,11 @@
 const webpack = require('webpack');
 let UglifyJsPlugin;
+let isUsingTerser = false;
+
+try {
```
Comment (GGAlanSmithee) [inline src/misc/get-uglify-plugin.js]:
```js
try {
  UglifyJsPlugin = require('terser-webpack-plugin');
  isUsingTerser = true;
} catch (e) {}

try {
  UglifyJsPlugin = webpack.optimize.UglifyJsPlugin;
} catch (e) {}

if (!UglifyJsPlugin) {
  try {
    UglifyJsPlugin = require('uglifyjs-webpack-plugin');
  } catch (e) {}
}
```

Regarding this logic, depending on what optimization libs are present, `terser-webpack-plugin` will not be used. For instance, I do not know under what circumstances `webpack.optimize.UglifyJsPlugin` will be available, but given the current implementation, `terser-webpack-plugin` will have the least presidence, is this correct? Maybe explore a different solution, maybe something like:

```js
try {
  UglifyJsPlugin = require('terser-webpack-plugin');
  isUsingTerser = true;

  if (!UglifyJsPlugin) {
    UglifyJsPlugin = webpack.optimize.UglifyJsPlugin;
  }

  if (!UglifyJsPlugin) {
    UglifyJsPlugin = require('uglifyjs-webpack-plugin');
  }
} catch (e) {}
```

?

Alternatively, wrap this in a function with early exits. What do you think?

File: src/misc/get-uglify-plugin.js [hunk 6]
```diff
@@ -11,4 +17,21 @@ if (!UglifyJsPlugin) {
   } catch (e) {}
 }
 
-module.exports = UglifyJsPlugin;
\ No newline at end of file
+module.exports = UglifyJsPlugin ? {
+  makeUglifyJsPlugin: (args) => {
```
[the remark this answers is not in the document]
Comment (GGAlanSmithee) [reply src/misc/get-uglify-plugin.js]: Yes, this is not possible to my knowledge either, don't see any other obvious choice.

**Comment by GGAlanSmithee** [comment]
@NekR do you have time to check this? Will require a new version and npm deploy!

**Comment by Bobgy** [comment]
@GGAlanSmithee Thank you for the detailed CR. I will address the issues.

File: tests/legacy/fixtures/sw-minify-false/webpack.config.js (hunk 1 above)
Comment (Bobgy) [reply tests/legacy/fixtures/sw-minify-false/webpack.config.js]: does look clearer

**Comment by Bobgy** [comment]
it seems CI is failing on the same code that passed before. I don't know why. I will debug soon

**Comment by GGAlanSmithee** [comment]
@Bobgy Thanks for your continued work. I see that you reverted your last commit, so I guess there is still something you want to address? Will await further comments from you.

**Comment by Bobgy** [comment]
@GGAlanSmithee I made the revert just to retry CI and it was still failing. It seemed that latest webpack was broken at that time. Now I just removed the revert and CI is passing again. Please review if you have time.

It seems I messed up a little bit when rebasing. Fixing it right now.

**Comment by Bobgy** [comment]
OK, all issues fixed and CI is passing again

File: src/service-worker.js [hunk 7]
```diff
@@ -156,13 +156,13 @@ export default class ServiceWorker {
     if (typeof this.minify === 'boolean') {
       minify = this.minify;
     } else {
-      minify = !!UglifyJsPlugin && (
+      minify = makeUglifyJsPlugin && (
         !!(
           compiler.options.optimization &&
           compiler.options.optimization.minimize
         ) || !!(
           (compiler.options.plugins || []).some((plugin) => {
```
Comment (GGAlanSmithee) [inline src/service-worker.js]:
```js
(compiler.options.plugins || []).some((plugin) => {
    return isInstanceOfUglifyJsPlugin(plugin);
})
```

can be shortened to

```js
(compiler.options.plugins || []).some(isInstanceOfUglifyJsPlugin)
```

File: docs/FAQ.md [hunk 8]
```diff
@@ -3,7 +3,7 @@
 ___________________________________
 
 **Is it possible to minify `ServiceWorker` script output?**  
-Yes, `offline-plugin` perfectly works with official `webpack.optimize.UglifyJsPlugin`, so if it's used you will get minified `ServiceWorker` script as well (no additional options required).
+Yes, `offline-plugin` perfectly works with official `webpack.optimize.UglifyJsPlugin` and `terser-webpack-plugin`, so if it's used you will get minified `ServiceWorker` script as well (no additional options required).
```
[the remark this answers is not in the document]
Comment (Bobgy) [reply docs/FAQ.md]: addressed

File: lib/misc/get-uglify-plugin.js [hunk 9]
```diff
@@ -2,15 +2,45 @@
 
 var webpack = require('webpack');
 var UglifyJsPlugin = undefined;
+var isUsingTerser = false;
 
-try {
-  UglifyJsPlugin = webpack.optimize.UglifyJsPlugin;
-} catch (e) {}
+if (!UglifyJsPlugin) {
```
[the remark this answers is not in the document]
Comment (Bobgy) [reply lib/misc/get-uglify-plugin.js]: Yes, I understand it's unnecessary. However, I think using the same format makes the pattern here more obvious -- the first found plugin for minification will be used. If we had the extra if before, I wouldn't have made the mistake of getting terser-webpack-plugin with a wrong priority.

[the remark this answers is not in the document]
Comment (Bobgy) [reply lib/misc/get-uglify-plugin.js]: what do you think?

File: src/misc/get-uglify-plugin.js [hunk 10]
```diff
@@ -1,14 +1,40 @@
 const webpack = require('webpack');
```
[the remark this answers is not in the document]
Comment (Bobgy) [reply src/misc/get-uglify-plugin.js]: totally makes sense, I will make the change

File: src/service-worker.js [hunk 11]
```diff
@@ -82,28 +82,28 @@ export default class ServiceWorker {
...
         test: new RegExp(name),
-        uglifyOptions: uglifyOptions
+        uglifyOptions,
       };
 
-      new UglifyJsPlugin(options).apply(childCompiler);
+      makeUglifyJsPlugin(options).apply(childCompiler);
     } else if (
-      (this.minify !== false || optimization.minimize) && UglifyJsPlugin
+      (this.minify !== false || optimization.minimize) && makeUglifyJsPlugin
```
[the remark this answers is not in the document]
Comment (Bobgy) [reply src/service-worker.js]: I'm not sure about the change. For example. `undefined === true` is false, but `undefined !== false` is true. I think the code here is dealing with several corner cases in a less obvious way.

Also that, I am making a change to this line, but only the last part. I don't think I should touch code that's irrelevant to my change.

File: src/service-worker.js (hunk 7 above)
Comment (Bobgy) [reply src/service-worker.js]: awesome, looks better

**Comment by Bobgy** [comment]
@GGAlanSmithee Thank you for your thoughtful comments.

**Comment by GGAlanSmithee** [comment]
image

So green 💯

**Pull request merged**